src/workspacer.Sound/SoundManager.cs

```diff
@@ -0,0 +1,393 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Vannatech.CoreAudio.Constants;
+using Vannatech.CoreAudio.Enumerations;
+using Vannatech.CoreAudio.Externals;
+using Vannatech.CoreAudio.Interfaces;
+
+namespace workspacer.Sound
+{
+    public delegate void DeviceEventDelegate(DeviceInfo device);
+    public delegate void DeviceDataEventDelegate<DataT>(DeviceInfo device, DataT newState);
+
+    public class SoundManager : IMMNotificationClient
+    {
+        private IMMDeviceEnumerator _mmDeviceEnumerator;
+
+        private List<SoundClient> _clients = new();
+
+        public event DeviceEventDelegate DeviceRemoved;
+        public event DeviceEventDelegate DeviceAdded;
+        public event DeviceDataEventDelegate<PROPERTYKEY> DevicePropertyValueChanged;
+        public event DeviceDataEventDelegate<uint> DeviceChanged;
+
+        public event DeviceDataEventDelegate<float> VolumeChanged;
+        public event DeviceDataEventDelegate<bool> MuteChanged;
+
+        public SoundManager()
+        {
+            _mmDeviceEnumerator = Activator.CreateInstance(Type.GetTypeFromCLSID(new Guid(ComCLSIDs.MMDeviceEnumeratorCLSID))) as IMMDeviceEnumerator;
+            if (_mmDeviceEnumerator == null)
+            {
+                throw new Exception("Creating device enumerator failed");
+            }
+
+            _mmDeviceEnumerator.RegisterEndpointNotificationCallback(this);
+        }
+
+        public SoundClient RegisterClientForDefault(DeviceType type, params DeviceRole[] roles)
+        {
+            var deviceId = GetDefaultDeviceId(type, roles);
+            var client = RegisterClient(deviceId, type, roles);
+
+            return client;
+        }
+
+        public SoundClient RegisterClientForDeviceId(string deviceId)
+        {
+            return RegisterClient(deviceId);
+        }
+
+        private SoundClient RegisterClient(string deviceId, DeviceType? type = null, params DeviceRole[] roles)
```

Comment:
Semantically `RegisterClient(...)` very much sounds like you as a caller are expected to provide a client of some sort... And it also sounds like something you may have to "unregister" later. But from what I see, there's no "unregister" API, so currently it feels like it's somewhat lacking a symmetry of sorts.

But given that we `RegisterClient(...)` don't expect a client at all, but rather *returns a "client" to the caller*, I think a better naming-convention for these methods would be `GetClientForXXX(...)`. How about it?